src/Servers/Kestrel/Core/src/Internal/Http2/Http2FrameWriter.cs

```diff
@@ -72,6 +77,188 @@ internal class Http2FrameWriter
+                }
+
+                if (producer.IsTimingWrite)
+                {
+                    _timeoutControl.StopTimingWrite();
+                }
+
+                producer.WriteHeaders = false;
+
+                reader.AdvanceTo(buffer.End);
+
+                if ((readResult.IsCompleted && !hasMoreData) || readResult.IsCanceled)
+                {
+                    await reader.CompleteAsync();
+
+                    producer.CompleteResponse();
+                }
+                // We're not going to schedule this again if there's no remaining window.
+                // When the window update is sent, the producer will be re-queued if needed.
+                else if (hasMoreData)
+                {
+                    // We have no more connection window, put this producer in a queue waiting for it to
+                    // a window update to resume the connection.
+                    if (remainingConnection == 0)
+                    {
+                        // Mark the output as waiting for a window upate to resume writing (there's still data)
+                        producer.MarkWaitingForWindowUpdates(true);
+
+                        lock (_windowUpdateLock)
+                        {
+                            _waitingForMoreWindow.Enqueue(producer);
+                        }
+
+                        // Include waiting for window updates in timing writes
+                        if (_minResponseDataRate != null)
+                        {
+                            producer.IsTimingWrite = true;
+                            _timeoutControl.StartTimingWrite();
+                        }
+                    }
+                    else if (remainingStream > 0)
+                    {
+                        // Move this stream to the back of the queue so we're being fair to the other streams that have data
+                        Schedule(producer);
+                    }
+                    else
+                    {
+                        // Mark the output as waiting for a window upate to resume writing (there's still data)
+                        producer.MarkWaitingForWindowUpdates(true);
+
+                        // Include waiting for window updates in timing writes
+                        if (_minResponseDataRate != null)
+                        {
+                            producer.IsTimingWrite = true;
+                            _timeoutControl.StartTimingWrite();
+                        }
+                    }
+                }
+                else if (reschedule)
+                {
+                    Schedule(producer);
+                }
+            }
+            catch (Exception ex)
+            {
+                _log.LogCritical(ex, "The event loop in connection {ConnectionId} failed unexpectedly", _connectionId);
+            }
+        }
+
+        _log.LogDebug("The connection processing loop for {ConnectionId} ended gracefully", _connectionId);
```

Comment:
This is logged for every completed connection. Should use defined log via source gen for perf.

Reply:
Yep, I need to clean that up.